### ISSUE TITLE
x64 branch table.insert

### ISSUE THREAD

**Issue by Shaar-games** [Facepunch/garrysmod-issues#4108]
table.insert with a huge number as key freeze the game ( work with no lag in x86 )

table.insert( {} , math.huge , "hi" )

**Comment by thegrb93** [comment]
2^31-1 is the highest safe value. Anything larger than 2^31 freezes.

**Comment by Verideth** [comment]
That's how GMods architecture works. It's x86.

By the way, `math.huge` is +infinity. You can't even do this. You're practically just making a massive table of infinities. Which is like, a massive table of `1/0`'s.

**Comment by thegrb93** [comment]
No, that's not how gmod works. Explain why greater than 2^31-1 values freeze rather than 2^32-1 on x64 and not x86. If you know anything about integers, you'd know the bug its running into.

**Comment by Verideth** [comment]
Read:

https://stevedonovan.github.io/lua-5.1.4/ltable.c.html#unbound_search
https://www.quora.com/What-is-the-use-of-RAM-in-a-computer-system
https://en.wikipedia.org/wiki/Undefined_behavior
https://twolivesleft.com/Codea/Reference/Lua.html#table.insert
https://wiki.garrysmod.com/page/math/huge

Lua code is done on a "heap allocated stack." Versus the native machines stack. Also, Lua has built in intrinsic support with bitwise operations, that can convert x64 numbers between x86 numbers pretty efficiently. But if you knew anything about numbers, you would know that, right? /s

I'm not even sure why this works on x86, to be honest. Perhaps Lua has built in functionality to stop it from breaking?

**Comment by thegrb93** [comment]
>I'm not even sure why this works on x86, to be honest. Perhaps Lua has built in functionality to stop it from breaking?

Yeah, and the signedness bug is most likely bypassing it

**Comment by thegrb93** [comment]
The way you keep linking basic computer science shit is really pretentious too so stop it.

**Comment by Verideth** [comment]
> The way you keep linking basic computer science shit to me is really pretentious too so stop it.

My bad, I'm sorry. Wasn't trying to insult you or anything. And it's really not "basic." It's pretty advanced stuff truly, especially on the lower level, which is what this bug is from.

**Comment by thegrb93** [comment]
I'm not against being linked useful stuff, but those are unhelpful.

You can look and see in the table lib that it uses 32bit signed ints, and my guess is luaL_checkint is no longer clamped. If you look at `tinsert (lua_State *L)` you can see how that would lead to an infinite loop.

**Comment by Verideth** [comment]
Not helpful? In what way? It pertains to the issue. And OP might want to know more.

That's the issue. If you check out and go through the call stack of these:
https://www.lua.org/source/5.1/lauxlib.c.html#luaL_checkinteger
https://www.lua.org/source/5.1/lapi.c.html#lua_tointeger
https://www.lua.org/source/5.1/lapi.c.html#index2adr

You'll notice in 'index2adr' 
`if (o >= L->top) return cast(TValue *, luaO_nilobject);`

It's returning a nil object if the index is greater than the "first _free slot_" of the stack. Which is out of memory bounds, when using `math.huge`. But then again, it still doesn't crash on x86 according to OP. But it's probably because Lua _generally_ has more support for x86.

**Comment by thegrb93** [comment]
>Not helpful? In what way? It pertains to the issue. And OP might want to know more.

![ram](https://user-images.githubusercontent.com/6713261/71777802-41ce3900-2f73-11ea-9dc3-c9288f4cc3b8.png)

**Comment by thegrb93** [comment]
Anyone who has useful input in this issue already knows everything you linked. You are providing no useful input into the discussion and (whether you know it or not) are rather being condescending and contradictory to anything against your own speculation. Admittedly I am being contradictory as well, but I'm at least not telling other developers to fucking read what the fuck ram is.

**Comment by RaphaelIT7** [comment]
Using math.huge doesn't really seem to be supported as in Lua 5.5 it would result in `bad argument #2 to 'insert' (number has no integer representation)`
For any **negative** number, LuaJIT begins to loop and tries to shift entries when there is nothing needed to be shifted, causing a massive performance issue with any large negative value.
This behavior **also exists in Lua 5.1** but in Lua 5.4 table.insert will throw an out of range error for any number below 1 and above INT_MAX.
I think the best solution would be to apply Lua 5.4's behavior and error for invalid numbers
Change: https://github.com/RaphaelIT7/gmod-holylib/commit/10e8ad31af05db9919549ec551e3c2d37fd00d04
**BUT** that would change existing behavior yet since using negative numbers definitely seems like undefined behavior, it probably would be good to do. (Though if Rubat wants a change that doesn't break existing behavior, I could look into that :))

Example how negative numbers are already kinda broken:
`lua_run local tab = {1, 2, 3} table.insert( tab, 0, "hi0") table.insert( tab, -1, "hi1") table.insert( tab, -1, "hi2" ) for k,v in pairs(tab) do print(k,v) end`
Result:
```txt
0       hi1
2       hi0 <--- It skipped 1?
4       1
5       2
6       3
-1      hi2
```

> [!NOTE]
> I've also tested this on `dev` branch and using `math.huge` also causes a freeze there, so this doesn't seem to be a 64x specific issue.

**Comment by artemking4** [comment]
It's very likely nobody is relying on this behavior, as it literally causes a huge as hell loop. Making this error out is a good idea, i second this

**Comment by RaphaelIT7** [comment]
There is a second possible implementation, to instead make it skip shifting as negative indexes are treated as hashed entries and not array entries.
Like if you already do `local tab = {} table.insert( tab, 1000, "hi" ) table.insert( tab, 1000, "hi2" )` the entry `1000` would be overwritten instead of being shifted since it's not an array entry, this same behavior can be applied for negative numbers by simply skipping the loop.
Change: https://github.com/RaphaelIT7/gmod-holylib/commit/fcd6751ac48ae29b4875acd31e356807653b8db8

Though personally I think throwing an error is better yet it would change how it would behave, this solution would purely solve the performance issue and prevent shifting for negative entries.
